File: server/src/routes/tasks.py
from flask import Blueprint, request, jsonify
from bson.objectid import ObjectId
import datetime, os
from pymongo import MongoClient
from .auth import authenticate_token

bp = Blueprint('tasks', __name__, url_prefix='/api/v1/tasks')
MONGODB_URI = os.getenv('MONGODB_URI', 'mongodb://localhost:27017/rezero')
client = MongoClient(MONGODB_URI)
db = client.get_default_database()
tasks = db['tasks']
ingests = db['ingests']
agent_jobs = db['agentjobs']
agent_results = db['agentresults']
audit_logs = db['auditlogs']

# Create new task
@bp.route('/', methods=['POST'])
@authenticate_token
def create_task():
    data = request.json
    ingest_id = data.get('ingestId')
    name = data.get('name')
    selected_agents = data.get('selectedAgents')
    description = data.get('description')
    parameters = data.get('parameters', {})
    priority = data.get('priority', 'medium')
    if not ingest_id or not name or not selected_agents:
        return jsonify({'error': 'Missing required fields'}), 400
    ingest = ingests.find_one({'_id': ObjectId(ingest_id), 'userId': request.user['_id'], 'status': 'completed'})
    if not ingest:
        return jsonify({'error': 'Ingest not found or not ready for processing'}), 404
    task = {
        'userId': request.user['_id'],
        'ingestId': ObjectId(ingest_id),
        'name': name,
        'description': description,
        'selectedAgents': selected_agents,
        'parameters': parameters,
        'priority': priority,
        'status': 'pending',
        'createdAt': datetime.datetime.utcnow(),
        'updatedAt': datetime.datetime.utcnow()
    }
    task_id = tasks.insert_one(task).inserted_id
    audit_logs.insert_one({
        'userId': request.user['_id'],
        'action': 'task_create',
        'resourceType': 'task',
        'resourceId': task_id,
        'ipAddress': request.remote_addr,
        'userAgent': request.headers.get('User-Agent'),
        'details': {'ingestId': ingest_id, 'selectedAgents': selected_agents, 'priority': priority},
        'status': 'success',
        'timestamp': datetime.datetime.utcnow()
    })
    return jsonify({'message': 'Task created successfully', 'taskId': str(task_id), 'status': 'pending'})

# Get task by ID
@bp.route('/<id>', methods=['GET'])
@authenticate_token
def get_task(id):
    task = tasks.find_one({'_id': ObjectId(id), 'userId': request.user['_id']})
    if not task:
        return jsonify({'error': 'Task not found'}), 404
    ingest = ingests.find_one({'_id': task['ingestId']})
    jobs = list(agent_jobs.find({'taskId': ObjectId(id)}))
    results = list(agent_results.find({'taskId': ObjectId(id)}))
    task['id'] = str(task['_id'])
    del task['_id']
    return jsonify({
        'task': task,
        'ingest': ingest,
        'agentJobs': jobs,
        'agentResults': results
    })

# List user's tasks
@bp.route('/', methods=['GET'])
@authenticate_token
def list_tasks():
    page = int(request.args.get('page', 1))
    limit = int(request.args.get('limit', 10))
    skip = (page - 1) * limit
    status = request.args.get('status')
    priority = request.args.get('priority')
    agent_type = request.args.get('agentType')
    sort_by = request.args.get('sortBy', 'createdAt')
    sort_order = -1 if request.args.get('sortOrder', 'desc') == 'desc' else 1
    filter = {'userId': request.user['_id']}
    if status:
        filter['status'] = status
    if priority:
        filter['priority'] = priority
    if agent_type:
        filter['selectedAgents'] = agent_type
    cursor = tasks.find(filter).sort(sort_by, sort_order).skip(skip).limit(limit)
    total = tasks.count_documents(filter)
    tasks_list = []
    for task in cursor:
        task['id'] = str(task['_id'])
        del task['_id']
        tasks_list.append(task)
    return jsonify({
        'tasks': tasks_list,
        'pagination': {
            'page': page,
            'limit': limit,
            'total': total,
            'pages': (total + limit - 1) // limit
        }
    })

# Update task
@bp.route('/<id>', methods=['PUT'])
@authenticate_token
def update_task(id):
    data = request.json
    updates = {}
    if 'name' in data:
        updates['name'] = data['name']
    if 'description' in data:
        updates['description'] = data['description']
    if 'priority' in data:
        updates['priority'] = data['priority']
    task = tasks.find_one({'_id': ObjectId(id), 'userId': request.user['_id']})
    if not task:
        return jsonify({'error': 'Task not found'}), 404
    if task['status'] != 'pending':
        return jsonify({'error': 'Can only update pending tasks'}), 400
    tasks.update_one({'_id': ObjectId(id)}, {'$set': updates})
    audit_logs.insert_one({
        'userId': request.user['_id'],
        'action': 'task_update',
        'resourceType': 'task',
        'resourceId': id,
        'ipAddress': request.remote_addr,
        'userAgent': request.headers.get('User-Agent'),
        'details': updates,
        'status': 'success',
        'timestamp': datetime.datetime.utcnow()
    })
    return jsonify({'message': 'Task updated successfully'})

# Cancel task
@bp.route('/<id>/cancel', methods=['POST'])
@authenticate_token
def cancel_task(id):
    task = tasks.find_one({'_id': ObjectId(id), 'userId': request.user['_id']})
    if not task:
        return jsonify({'error': 'Task not found'}), 404
    if task['status'] not in ['pending', 'running']:
        return jsonify({'error': 'Can only cancel pending or running tasks'}), 400
    tasks.update_one({'_id': ObjectId(id)}, {'$set': {'status': 'cancelled', 'completedAt': datetime.datetime.utcnow()}})
    agent_jobs.update_many({'taskId': ObjectId(id), 'status': {'$in': ['queued', 'running']}}, {'$set': {'status': 'cancelled', 'completedAt': datetime.datetime.utcnow()}})
    audit_logs.insert_one({
        'userId': request.user['_id'],
        'action': 'task_cancel',
        'resourceType': 'task',
        'resourceId': id,
        'ipAddress': request.remote_addr,
        'userAgent': request.headers.get('User-Agent'),
        'status': 'success',
        'timestamp': datetime.datetime.utcnow()
    })
    return jsonify({'message': 'Task cancelled successfully'})

# Delete task
@bp.route('/<id>', methods=['DELETE'])
@authenticate_token
def delete_task(id):
    task = tasks.find_one({'_id': ObjectId(id), 'userId': request.user['_id']})
    if not task:
        return jsonify({'error': 'Task not found'}), 404
    agent_jobs.delete_many({'taskId': ObjectId(id)})
    agent_results.delete_many({'taskId': ObjectId(id)})
    tasks.delete_one({'_id': ObjectId(id)})
    audit_logs.insert_one({
        'userId': request.user['_id'],
        'action': 'task_delete',
        'resourceType': 'task',
        'resourceId': id,
        'ipAddress': request.remote_addr,
        'userAgent': request.headers.get('User-Agent'),
        'status': 'success',
        'timestamp': datetime.datetime.utcnow()
    })
    return jsonify({'message': 'Task deleted successfully'})

# Get available agents (stub)
@bp.route('/agents/available', methods=['GET'])
@authenticate_token
def available_agents():
    # This should be replaced with real agent discovery logic
    agents = [
        {'type': 'data_extraction', 'capabilities': {'agentType': 'Data Extraction Agent', 'inputTypes': ['text', 'json', 'csv']}},
        {'type': 'financial_analysis', 'capabilities': {'agentType': 'Financial Analysis Agent', 'inputTypes': ['text', 'json', 'csv']}},
        {'type': 'news_summarization', 'capabilities': {'agentType': 'News Summarization Agent', 'inputTypes': ['text', 'json', 'csv']}}
    ]
    return jsonify({'agents': agents})

File: server/src/routes/tasks.js
import express from 'express';
import { body, validationResult } from 'express-validator';
import Task from '../models/Task.js';
import Ingest from '../models/Ingest.js';
import AgentJob from '../models/AgentJob.js';
import AgentResult from '../models/AgentResult.js';
import AuditLog from '../models/AuditLog.js';
import { MetaAgent } from '../agents/MetaAgent.js';
import { authenticateToken } from './auth.js';
import logger from '../utils/logger.js';

const router = express.Router();
const metaAgent = new MetaAgent();

// Create new task
router.post('/', authenticateToken, [
  body('ingestId').isMongoId(),
  body('name').trim().isLength({ min: 1 }),
  body('selectedAgents').isArray({ min: 1 }),
  body('description').optional().trim(),
  body('parameters').optional().isObject()
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { 
      ingestId, 
      name, 
      selectedAgents, 
      description, 
      parameters = {},
      priority = 'medium'
    } = req.body;

    // Validate ingest exists and belongs to user
    const ingest = await Ingest.findOne({
      _id: ingestId,
      userId: req.user._id,
      status: 'completed'
    });

    if (!ingest) {
      return res.status(404).json({ 
        error: 'Ingest not found or not ready for processing' 
      });
    }

    // Validate agent selection
    try {
      metaAgent.validateAgentSelection(selectedAgents);
    } catch (error) {
      return res.status(400).json({ error: error.message });
    }

    // Create task
    const task = new Task({
      userId: req.user._id,
      ingestId,
      name,
      description,
      selectedAgents,
      parameters: new Map(Object.entries(parameters)),
      priority
    });

    await task.save();

    // Log task creation
    await AuditLog.logAction({
      userId: req.user._id,
      action: 'task_create',
      resourceType: 'task',
      resourceId: task._id,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
      details: { 
        ingestId, 
        selectedAgents, 
        priority 
      },
      status: 'success'
    });

    logger.info('Task created', { 
      taskId: task._id, 
      userId: req.user._id, 
      selectedAgents 
    });

    // Start task processing asynchronously
    setImmediate(async () => {
      try {
        logger.info('Starting async task processing', { 
          taskId: task._id, 
          selectedAgents,
          inputLength: ingest.processedContent?.length || 0
        });
        
        const result = await metaAgent.orchestrateTask(task, ingest.processedContent, selectedAgents, parameters);
        if (result.status === 'completed') {
          logger.info('Task completed successfully', { 
            taskId: task._id, 
            duration: result.metadata.duration 
          });
        } else {
          logger.error('Task failed', { 
            taskId: task._id, 
            error: result.error 
          });
        }
      } catch (error) {
        logger.error('Task orchestration error', { 
          taskId: task._id, 
          error: error.message,
          stack: error.stack
        });
        try {
          await task.fail(error);
        } catch (failError) {
          logger.error('Failed to update task status to failed', {
            taskId: task._id,
            error: failError.message
          });
        }
      }
    });

    res.status(201).json({
      message: 'Task created successfully',
      taskId: task._id,
      status: 'pending'
    });

  } catch (error) {
    logger.error('Task creation failed', { error: error.message });
    res.status(500).json({ error: 'Failed to create task' });
  }
});

// Get task by ID
router.get('/:id', authenticateToken, async (req, res) => {
  try {
    const { id } = req.params;

    const task = await Task.findOne({
      _id: id,
      userId: req.user._id
    }).populate('ingestId', 'type status metadata');

    if (!task) {
      return res.status(404).json({ error: 'Task not found' });
    }

    // Get agent jobs for this task
    const agentJobs = await AgentJob.find({ taskId: id })
      .sort({ createdAt: 1 });

    // Get agent results for this task
    const agentResults = await AgentResult.find({ taskId: id })
      .sort({ createdAt: 1 });

    res.json({
      task: {
        id: task._id,
        name: task.name,
        description: task.description,
        status: task.status,
        selectedAgents: task.selectedAgents,
        parameters: Object.fromEntries(task.parameters),
        priority: task.priority,
        progress: task.progress,
        startedAt: task.startedAt,
        completedAt: task.completedAt,
        estimatedDuration: task.estimatedDuration,
        actualDuration: task.actualDuration,
        error: task.error,
        metadata: Object.fromEntries(task.metadata),
        ingest: task.ingestId,
        createdAt: task.createdAt,
        updatedAt: task.updatedAt
      },
      agentJobs: agentJobs.map(job => ({
        id: job._id,
        agentType: job.agentType,
        status: job.status,
        priority: job.priority,
        retryCount: job.retryCount,
        startedAt: job.startedAt,
        completedAt: job.completedAt,
        estimatedDuration: job.estimatedDuration,
        actualDuration: job.actualDuration,
        error: job.error,
        metadata: Object.fromEntries(job.metadata),
        createdAt: job.createdAt,
        updatedAt: job.updatedAt
      })),
      agentResults: agentResults.map(result => ({
        id: result._id,
        agentType: result.agentType,
        resultType: result.resultType,
        title: result.title,
        content: result.content,
        structuredData: Object.fromEntries(result.structuredData),
        confidence: result.confidence,
        quality: result.quality,
        tags: result.tags,
        metadata: Object.fromEntries(result.metadata),
        provenance: result.provenance,
        isAggregated: result.isAggregated,
        createdAt: result.createdAt,
        updatedAt: result.updatedAt
      }))
    });

  } catch (error) {
    logger.error('Get task failed', { error: error.message });
    res.status(500).json({ error: 'Failed to get task' });
  }
});

// List user's tasks
router.get('/', authenticateToken, async (req, res) => {
  try {
    const { 
      page = 1, 
      limit = 10, 
      status, 
      priority, 
      agentType,
      sortBy = 'createdAt',
      sortOrder = 'desc'
    } = req.query;
    
    const skip = (page - 1) * limit;
    const sort = { [sortBy]: sortOrder === 'desc' ? -1 : 1 };

    const filter = { userId: req.user._id };
    if (status) filter.status = status;
    if (priority) filter.priority = priority;
    if (agentType) filter.selectedAgents = agentType;

    const tasks = await Task.find(filter)
      .populate('ingestId', 'type status metadata')
      .sort(sort)
      .skip(skip)
      .limit(parseInt(limit));

    const total = await Task.countDocuments(filter);

    res.json({
      tasks: tasks.map(task => ({
        id: task._id,
        name: task.name,
        description: task.description,
        status: task.status,
        selectedAgents: task.selectedAgents,
        priority: task.priority,
        progress: task.progress,
        startedAt: task.startedAt,
        completedAt: task.completedAt,
        actualDuration: task.actualDuration,
        error: task.error,
        ingest: task.ingestId,
        createdAt: task.createdAt,
        updatedAt: task.updatedAt
      })),
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total,
        pages: Math.ceil(total / limit)
      }
    });

  } catch (error) {
    logger.error('List tasks failed', { error: error.message });
    res.status(500).json({ error: 'Failed to list tasks' });
  }
});

// Update task
router.put('/:id', authenticateToken, [
  body('name').optional().trim().isLength({ min: 1 }),
  body('description').optional().trim(),
  body('priority').optional().isIn(['low', 'medium', 'high'])
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { id } = req.params;
    const { name, description, priority } = req.body;

    const task = await Task.findOne({
      _id: id,
      userId: req.user._id
    });

    if (!task) {
      return res.status(404).json({ error: 'Task not found' });
    }

    // Only allow updates for pending tasks
    if (task.status !== 'pending') {
      return res.status(400).json({ 
        error: 'Can only update pending tasks' 
      });
    }

    const updates = {};
    if (name) updates.name = name;
    if (description !== undefined) updates.description = description;
    if (priority) updates.priority = priority;

    const updatedTask = await Task.findByIdAndUpdate(
      id,
      updates,
      { new: true, runValidators: true }
    );

    // Log task update
    await AuditLog.logAction({
      userId: req.user._id,
      action: 'task_update',
      resourceType: 'task',
      resourceId: id,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
      details: updates,
      status: 'success'
    });

    logger.info('Task updated', { taskId: id, userId: req.user._id });

    res.json({
      message: 'Task updated successfully',
      task: {
        id: updatedTask._id,
        name: updatedTask.name,
        description: updatedTask.description,
        status: updatedTask.status,
        priority: updatedTask.priority,
        updatedAt: updatedTask.updatedAt
      }
    });

  } catch (error) {
    logger.error('Task update failed', { error: error.message });
    res.status(500).json({ error: 'Failed to update task' });
  }
});

// Cancel task
router.post('/:id/cancel', authenticateToken, async (req, res) => {
  try {
    const { id } = req.params;

    const task = await Task.findOne({
      _id: id,
      userId: req.user._id
    });

    if (!task) {
      return res.status(404).json({ error: 'Task not found' });
    }

    // Only allow cancellation of pending or running tasks
    if (!['pending', 'running'].includes(task.status)) {
      return res.status(400).json({ 
        error: 'Can only cancel pending or running tasks' 
      });
    }

    task.status = 'cancelled';
    task.completedAt = new Date();
    task.actualDuration = task.duration;
    await task.save();

    // Cancel associated agent jobs
    await AgentJob.updateMany(
      { taskId: id, status: { $in: ['queued', 'running'] } },
      { status: 'cancelled', completedAt: new Date() }
    );

    // Log task cancellation
    await AuditLog.logAction({
      userId: req.user._id,
      action: 'task_cancel',
      resourceType: 'task',
      resourceId: id,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
      status: 'success'
    });

    logger.info('Task cancelled', { taskId: id, userId: req.user._id });

    res.json({
      message: 'Task cancelled successfully'
    });

  } catch (error) {
    logger.error('Task cancellation failed', { error: error.message });
    res.status(500).json({ error: 'Failed to cancel task' });
  }
});

// Delete task
router.delete('/:id', authenticateToken, async (req, res) => {
  try {
    const { id } = req.params;

    const task = await Task.findOne({
      _id: id,
      userId: req.user._id
    });

    if (!task) {
      return res.status(404).json({ error: 'Task not found' });
    }

    // Delete associated agent jobs and results
    await AgentJob.deleteMany({ taskId: id });
    await AgentResult.deleteMany({ taskId: id });
    await Task.findByIdAndDelete(id);

    // Log task deletion
    await AuditLog.logAction({
      userId: req.user._id,
      action: 'task_delete',
      resourceType: 'task',
      resourceId: id,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
      status: 'success'
    });

    logger.info('Task deleted', { taskId: id, userId: req.user._id });

    res.json({
      message: 'Task deleted successfully'
    });

  } catch (error) {
    logger.error('Task deletion failed', { error: error.message });
    res.status(500).json({ error: 'Failed to delete task' });
  }
});

// Get available agents
router.get('/agents/available', authenticateToken, async (req, res) => {
  try {
    const agents = metaAgent.getAvailableAgents();

    res.json({
      agents: agents.map(agent => ({
        type: agent.type,
        capabilities: agent.capabilities
      }))
    });

  } catch (error) {
    logger.error('Get available agents failed', { error: error.message });
    res.status(500).json({ error: 'Failed to get available agents' });
  }
});

export default router;
